### PR TITLE
[mosaic-gpu] add reduction support in copy_smem_to_gmem

### PR DESCRIPTION
**Pull request by Amir-19**
This PR adds support for reduction operation--only `add` at this moment for `fp` dtype.

- `element_type` is now required to be provided to `mosaic_gpu_init_tma_desc` since in the case of reduction the tma descriptor needs to know exactly what is the dtype. for regular tma `(cp.async.bulk)` this was not an issue but with reduction (cp.reduce.async.bulk), incorrect dtype would result in hangs, incorrect results, or illegal instruction. 

- `nvvm.cp_async_bulk_tensor_reduce` needs to be called by only one thread otherwise it would result in multiple reductions in the same slice, uniform_ctx has to account for reduction.

@nvcastet @apaszke

### CODE REVIEW

File: jaxlib/mosaic/gpu/runtime.cc [hunk 1]
```diff
@@ -22,7 +22,7 @@ limitations under the License.
 extern "C" {
 
 void mosaic_gpu_init_tma_desc(CUtensorMap *tma_desc, void *base_addr,
-                              int64_t elem_bitwidth, int64_t rank,
+                              int64_t elem_type, int64_t elem_bitwidth, int64_t rank,
```
Comment (apaszke) [inline jaxlib/mosaic/gpu/runtime.cc:25]: If we have the elem_type, we also have elem_bitwidth, right? There's no point in keeping both args. Just delete the bitwidth and compute it below

File: tests/pallas/mosaic_gpu_test.py [hunk 2]
```diff
@@ -347,6 +348,34 @@ def kernel(x_ref, o_ref_gmem, scratch_ref):
...
+  def test_copy_smem_to_gmem_reduction(self, num_blocks, dtype):
+    @functools.partial(
+    pl.pallas_call,
+    grid=(num_blocks,),
+    in_specs=[pl.BlockSpec((128,), lambda *i: i), pl.BlockSpec(memory_space=plgpu.GMEM)],
+    out_specs=pl.BlockSpec(memory_space=plgpu.GMEM),
+    out_shape=jax.ShapeDtypeStruct([128], dtype),
+    scratch_shapes=[plgpu.SMEM((128,), dtype)],
+    input_output_aliases={1:0}
+    )
```
Comment (apaszke) [inline tests/pallas/mosaic_gpu_test.py:365]: nit: can you please fix formatting? we usually indent the arguments to a call (by 4 spaces, but 2 is fine too)

File: jax/experimental/mosaic/gpu/utils.py [hunk 3]
```diff
@@ -1205,3 +1205,58 @@ def bitcast(x: ir.Value, new_type: ir.Type):
...
+  elif reduction_op == TMAReductionKind.DEC:
+    return nvvm.TMAReduxKind.DEC
+  elif reduction_op == TMAReductionKind.AND:
+    return nvvm.TMAReduxKind.AND
+  elif reduction_op == TMAReductionKind.OR:
+    return nvvm.TMAReduxKind.OR
+  elif reduction_op == TMAReductionKind.XOR:
+    return nvvm.TMAReduxKind.XOR
+  else:
+    raise ValueError(f"unsupported reduction kind {reduction_op}")
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/utils.py:1262]: It would make sense to make this a method of the enum above. Also this whole function can be simplified to `getattr(nvvm.TMAReduxKind, reduction_op.name)`

File: jax/experimental/mosaic/gpu/utils.py [hunk 4]
```diff
@@ -1205,3 +1205,58 @@ def bitcast(x: ir.Value, new_type: ir.Type):
 
 def ceil_div(x: int, y: int):
   return (x + y - 1) // y
+
+
+def get_tma_dtype(elem_type):
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/utils.py:1210]: This would make a good method of the enum below (once you move it to launch_context.py)

File: tests/pallas/mosaic_gpu_test.py [hunk 5]
```diff
@@ -347,6 +348,34 @@ def kernel(x_ref, o_ref_gmem, scratch_ref):
...
+    def kernel(x_ref, o_ref_gmem, o_ref_gmem_alias, scratch_ref):
+      del o_ref_gmem_alias
+      scratch_ref[...] = x_ref[...]
+      plgpu.commit_smem()
+      plgpu.copy_smem_to_gmem(scratch_ref.at[...], o_ref_gmem.at[...], reduction_op=mgpu_utils.TMAReductionKind.ADD)
```
Comment (apaszke) [inline tests/pallas/mosaic_gpu_test.py:370]: Wait, are TMA reductions guaranteed to be atomic? This is not discussed anywhere in the PTX ISA doc as far as I can tell

File: tests/pallas/mosaic_gpu_test.py [hunk 6]
```diff
@@ -347,6 +348,34 @@ def kernel(x_ref, o_ref_gmem, scratch_ref):
     x = jnp.arange(256).astype(jnp.float32)
     np.testing.assert_array_equal(kernel(x)[indexer], x[indexer] + 1.0)
 
+  @parameterized.named_parameters(
+      {"testcase_name": "4_blocks_f16", "num_blocks": 4, "dtype": 'float16'},
+      {"testcase_name": "4_blocks_f32", "num_blocks": 4, "dtype": 'float32'},
+      {"testcase_name": "2_blocks_bf16", "num_blocks": 2, "dtype": 'bfloat16'},
```
Comment (apaszke) [inline tests/pallas/mosaic_gpu_test.py:354]: There's no value in parameterizing this test by the number of blocks. Perhaps just hardcode 200 blocks or so (to at least use all SMs). It does make sense to parameterize the test by the dtype!

**Review by apaszke — COMMENTED** [review]
Very cool, thank you!

File: jax/_src/pallas/mosaic_gpu/primitives.py [hunk 7]
```diff
@@ -186,6 +188,7 @@ def copy_smem_to_gmem(
     predicate: jax.Array | None = None,
     *,
     commit_group: bool = True,
+    reduction_op: mgpu_utils.TMAReductionKind | None = None,
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/_src/pallas/mosaic_gpu/primitives.py]: I changed it to Literal[list of possible reduction ops]

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 8]
```diff
@@ -375,6 +376,7 @@ def async_copy(
       collective: Sequence[gpu.Dimension] | gpu.Dimension | None = None,
       partitioned: int | None = None,
       predicate: ir.Value | None = None,  # Should select 0 or 1 threads from the WG.
+      reduction_op: utils.TMAReductionKind | None = None,
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: good point! moved to `launch_context.py`

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 9]
```diff
@@ -607,7 +614,7 @@ def partition_dim(dim: int, idx: ir.Value, num_chunks: int):
 
     uniform_ctx = (
         functools.partial(utils.single_thread, per_block=False)
-        if uniform and predicate is None
+        if uniform and (predicate is None or reduction_op is not None)
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: I removed this but had to change the context for reduction to `with utils.single_thread(per_block=False):` otherwise each thread would perform the reduction and we end up with `element=desired_value*num_threads` instead of `desired_value`. please let me know if you know of a better approach.

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 10]
```diff
@@ -679,12 +686,22 @@ def partition_dim(dim: int, idx: ir.Value, num_chunks: int):
           )
     else:
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py:741]: good catch! I believe you mean gmem->smem branch and I added the assertion.

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 11]
```diff
@@ -679,12 +686,22 @@ def partition_dim(dim: int, idx: ir.Value, num_chunks: int):
...
-        nvvm.cp_async_bulk_tensor_global_shared_cta(
-            tma_desc, smem_ptr, rev_dyn_base_indices, predicate=predicate
-        )
-        if arrive:
-          nvvm.cp_async_bulk_commit_group()
+      if reduction_op is not None:
+        with uniform_ctx():
+          reduction_kind = utils.get_tma_reduction_kind(reduction_op)
+          # by default, mode is nvvm.TMAStoreMode.TILE
+          nvvm.cp_async_bulk_tensor_reduce(
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py:747]: could you please elaborate how would you like predicate to be implemented here. I'm not sure if current change from `with uniform_ctx():` to `with utils.single_thread(per_block=False)`, but otherwise it wouldn't work as I mentioned above.

File: jax/experimental/mosaic/gpu/utils.py [hunk 12]
```diff
@@ -1205,3 +1205,58 @@ def bitcast(x: ir.Value, new_type: ir.Type):
...
+    elif bitwidth == 64:
+      return 4 # uint64
+  elif ir.F16Type.isinstance(elem_type):
+    return 6
+  elif ir.F32Type.isinstance(elem_type):
+    return 7
+  elif ir.BF16Type.isinstance(elem_type):
+    return 9
+  else:
+    return -1 # unsupported tma dtype
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/utils.py]: good point! done!

File: jaxlib/mosaic/gpu/runtime.cc (hunk 1 above)
Comment (Amir-19) [reply jaxlib/mosaic/gpu/runtime.cc]: added the logic to calculate `elem_bitwidth` based on `elem_type`

File: tests/pallas/mosaic_gpu_test.py (hunk 2 above)
Comment (Amir-19) [reply tests/pallas/mosaic_gpu_test.py:381]: done!

File: jax/experimental/mosaic/gpu/utils.py (hunk 3 above)
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/utils.py]: thank you! this looks much better now.

File: jax/experimental/mosaic/gpu/utils.py (hunk 4 above)
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/utils.py]: I think adding this to the enum is confusing as `get_tma_dtype` is used for non-reduction tma too and the enum represents the types of reduction for tma.

File: tests/pallas/mosaic_gpu_test.py (hunk 5 above)
Comment (Amir-19) [reply tests/pallas/mosaic_gpu_test.py]: yes. the updates to each individual element are atomic but it is not guaranteed that every single element updated by thread_i is touched before any of the updates from thread_j are performed. this is discussed [here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#:~:text=Each%20reduction%20operation%20performed%20by%20the) which mentions "Each reduction operation performed by the `cp.reduce.async.bulk` has individually `.relaxed.gpu` memory ordering semantics". I discussed with the team to add details about atomicity as it is more clear.

File: tests/pallas/mosaic_gpu_test.py (hunk 6 above)
Comment (Amir-19) [reply tests/pallas/mosaic_gpu_test.py]: done!

File: jax/experimental/mosaic/gpu/launch_context.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: I ended up using the ptx instruction instead of the nvvm call.

File: jax/experimental/mosaic/gpu/launch_context.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: after changing nvvm call to ptx instruction this is no longer required.

File: jax/experimental/mosaic/gpu/utils.py (hunk 4 above)
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/utils.py]: I removed the enum as I no longer use nvvm call for tma reduction.

File: jaxlib/mosaic/gpu/runtime.cc [hunk 13]
```diff
@@ -54,19 +93,6 @@ void mosaic_gpu_init_tma_desc(CUtensorMap *tma_desc, void *base_addr,
     elem_bytewidth = elem_bitwidth / 8;
```
Comment (apaszke) [inline jaxlib/mosaic/gpu/runtime.cc:93]: We override elem_bytewidth here, but you didn't update the `data_type`. We used to only select it below. Either way, I think it's enough if you simply use `CU_TENSOR_MAP_DATA_TYPE_UINT8` in the branch above

Comment (apaszke) [reply jaxlib/mosaic/gpu/runtime.cc:93]: Now when I read this, it looks like this PR took away our ability to TMA int4s, which the old code used to support. I think we might need to come up with our own enum of `elem_type` that diverges from the CUDA one, because we want to support more data types. Alternatively, we could normalize the int4s into uint8s on the Python side and keep this code simple.

**Review by apaszke — COMMENTED** [review]
Mostly LGTM, but I noticed the TMA descriptor creation now does not account for 4-bit types that we need to support

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 14]
```diff
@@ -229,6 +229,54 @@ def batch(self, leading_rank: int) -> MemRefTransform:
...
+
+  @staticmethod
+  def get_tma_reduction_kind(reduction_op: 'TMAReductionKind'):
+    try:
+      return getattr(nvvm.TMAReduxKind, reduction_op.name)
+    except AttributeError:
+      raise ValueError(f"unsupported reduction kind {reduction_op}")
+
+  @staticmethod
+  def string_to_reduction_kind(s: str):
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: moved this inside _get_tma_desc method.

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 15]
```diff
@@ -229,6 +229,54 @@ def batch(self, leading_rank: int) -> MemRefTransform:
...
+  if isinstance(elem_type, ir.IntegerType):
+    bitwidth = utils.bitwidth_impl(elem_type)
+    if bitwidth == 8:
+      return 0 # uint8
+    elif bitwidth == 16:
+      return 1 # uint16
+    elif bitwidth == 32:
+      return 2 # uint32
+    elif bitwidth == 64:
+      return 4 # uint64
```
[the remark this answers is not in the document]
Comment (Amir-19) [reply jax/experimental/mosaic/gpu/launch_context.py]: since this logic is now moved inside `_get_tma_desc` method, had to change `_get_tma_desc` signiture to get `reduction_op` arg.

File: jaxlib/mosaic/gpu/runtime.cc (hunk 13 above)
Comment (Amir-19) [reply jaxlib/mosaic/gpu/runtime.cc:98]: added int4s using the next available enum value 16.

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 16]
```diff
@@ -337,10 +340,38 @@ def init_tma_desc(host_ptr):
...
+          elif bitwidth == 4:
+            # NOTE: this dtype diverges from the supported types in cuda and is used
+            # to support int4
+            tma_dtype = 16
+        elif ir.F16Type.isinstance(ref_ty.element_type):
+          tma_dtype = 6
+        elif ir.F32Type.isinstance(ref_ty.element_type):
+          tma_dtype = 7
+        elif ir.BF16Type.isinstance(ref_ty.element_type):
+          tma_dtype = 9
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/launch_context.py:368]: Could we just assign them our own numbering that is separate from CUDA enums? I'm afraid that in the future TMA will be able to handle more types and we'll end up with 16 being used in the CUDA enum. If we just say this is our own convention it all becomes clearer

Comment (apaszke) [reply jax/experimental/mosaic/gpu/launch_context.py:368]: I think we should also add a check on `jaxlib.__version__` that says that if the version is less than 0.5.4 then we'll still pass in the bitwidth. Otherwise we'll use the enum you have here. If we don't do this, if someone installs an old jaxlib release (which JAX does support), we will end up calling the old runtime functions with the enum values in place of bitwidths.

**Review by apaszke — APPROVED** [review]
Nice! There's only one final snag around versioning of those changes, but this will be ready to submit as soon as that's fixed.

File: jax/experimental/mosaic/gpu/launch_context.py [hunk 17]
```diff
@@ -229,6 +231,13 @@ def batch(self, leading_rank: int) -> MemRefTransform:
 OnDeviceProfiler = profiler.OnDeviceProfiler
 
 
+def parse_jaxlib_version(v: str) -> tuple[int, ...]:
+  _version_regex = re.compile(r"([0-9]+(?:\.[0-9]+)*)(?:(rc|dev).*)?")
+  m = _version_regex.match(v)
+  if m is None:
+    raise ValueError(f"Unable to parse jaxlib version '{v}'")
+  return tuple(int(x) for x in m.group(1).split('.'))
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/launch_context.py:239]: I think it would be simpler to just do this:
```
from jax._src import lib as jaxlib
jaxlib.version  # already has type tuple[int, ...]
```

**Review by apaszke — APPROVED** [review]
LGTM but please use don't implement custom version parsing code

**Review by apaszke — APPROVED** [review]
Thanks!
